[WiFi] Reduce wifi reset calls and call full wifi init at reset (#3147)

TD-er opened the pull request:
Fixes: #2931
Fixes: #1258
Fixes: #3155
Fixes: #3141
Fixes: #3111 

Connected state is now based on the state reported by the internal WiFi object.
Frequent calls to `reset()` of WiFi are now rate limited to make sure we don't end up in a loop.

TD-er committed:
[WiFi] Reduce wifi reset calls and call full wifi init at reset

TD-er commented:
I made a test build for this.

- [ESPEasy_ESP32_mega-20200703-10-PR_3148.zip](https://www.dropbox.com/s/5si1dcq2xo4p4oz/ESPEasy_ESP32_mega-20200703-10-PR_3148.zip?dl=0)
- [ESPEasy_ESP82xx_mega-20200703-10-PR_3148.zip](https://www.dropbox.com/s/nhsshr2k288dvei/ESPEasy_ESP82xx_mega-20200703-10-PR_3148.zip?dl=0)

v-a-d-e-r commented:
I get often this now with this changes  (merged code, not the bin files above) and OTA update:
Update successful: 691952 bytes
Rebooting...
WIFI : WiFiConnected() out of sync
WIFI : Resetting...
Exception (28):
epc1=0x402773a1 epc2=0x00000000 epc3=0x00000000 excvaddr=0x00000039 depc=0x00000000
>>>stack>>>

TD-er commented:
I'm also testing this and there is some logic error which I'm hunting down as we speak.
Thanks for letting me know it isn't just my node.

P.S. to get the node to enter some endless loop, try to disconnect the ESP from the access point side (Mikrotik allows you to kick of a connected client)

v-a-d-e-r commented:
@TD-er: Something OT: Have you checked the pubSubClient v2.8? Sending works, receiving not... :-/

TD-er commented:
Nope, have not looked at the PubSubClient main repo for a while.

v-a-d-e-r commented:
I downloaded it here (right version?):
https://github.com/knolleary/pubsubclient/releases/tag/v2.8
Did you made some changes in the prev versions to work with ESPeasy?

TD-er commented:
> Did you made some changes in the prev versions to work with ESPeasy?

Yep.
Made quite a few changes, that's why Tasmota is also using the "ESPeasy version" of PubSubClient :)

v-a-d-e-r commented:
Ah, ok. So I can wait for a long time to get it to work in the orig condition... LOL

uzi18 commented:
Have a discussion with you some time ago in different issue, to prevent fast switching of wifi state - wifi logic should work more lazy (2nd step) and wifistate/wificonnected should be not a function but int/bool to check (first step), this will reduce wifi overhead.
But maybe it is only my point of view.

TD-er commented:
Not sure if I understand you correctly @uzi18 

What I'm now trying to do (not yet committed) is to consider the `wifiStatus` a proper bitfield.

What I'm seeing though is that this `wifiStatus` can get out of sync with the true state.
Also it really does happen (quite often) the WiFi reports to have an IP address set (via DHCP for example) and claims not to be connected.
It looks like I'm now narrowing in on the true cause of the crashes we sometimes see and which is very well reproducible when enabling syslog.
So far, it seems sending UDP packets while the WiFi is in this unknown state is causing these crashes.
As if it is queuing the packets and still processing this queue after the queue itself has been destructed.

TD-er committed:
[WiFi] Add timeout to get out of incorrect WiFi.connected() state

The WiFi.connected() call sometimes returns `WL_DISCONNECTED` even though it is connected and serving web pages etc.

Added a timeout of 15 seconds to get out of this or else the WiFi state is reset to try again fresh.

TD-er commented:
Made a new test build for this PR:

- [ESPEasy_ESP82xx_mega-20200703-12-PR_3148.zip](https://www.dropbox.com/s/htrylluq924fxzs/ESPEasy_ESP82xx_mega-20200703-12-PR_3148.zip?dl=0)
- [ESPEasy_ESP32_mega-20200703-12-PR_3148.zip](https://www.dropbox.com/s/98cpwqx73s7ahw7/ESPEasy_ESP32_mega-20200703-12-PR_3148.zip?dl=0)

uzi18 commented on src/ESPEasyWiFiEvent.cpp:
```diff
@@ -39,7 +39,10 @@ void setUseStaticIP(bool enabled) {
+  // Create the 'got IP event' so mark the wifiStatus to not have the got IP flag set
+  // This also implies the services are not fully initialized.
   wifiStatus     &= ~ESPEASY_WIFI_GOT_IP;
+  wifiStatus     &= ~ESPEASY_WIFI_SERVICES_INITIALIZED;
```
@TD-er use bitRead/bitWrite macros, do you use here inversed logic?

TD-er replied:
That's perhaps indeed _a bit_ more readable :)

TD-er committed:
[WiFi] Convert bit-wise operations to use bitSet/bitClear/bitRead

TD-er commented:
OK, there is really something wrong in the core library.
The work-around I'm working on is in short something like this:
```c++
station_status_t status = wifi_station_get_connect_status();
if (status == STATION_CONNECTING && timeOutReached(last_wifi_connect_attempt_moment + 15000)) {
      resetWiFi();
}
```

The strange part is, this `STATION_CONNECTING` is not being checked for in the `wl_status_t ESP8266WiFiSTAClass::status()` function.

But it is actually happening on only some modules when trying to reconnect, or on the first connection.
Exactly as has been reported for ages but I was unable to reproduce it until a module I just yesterday opened from its seal.

v-a-d-e-r commented:
After merging this: 
"[WiFi] Add timeout to get out of incorrect WiFi.connected() state"
the shown Vcc value is wrong (normal 3V, now 1.58V)? What has the WiFi stuff to do with Vcc? LC droped also a lot...
Can you check this please?

TD-er commented:
Is Vcc only dropped on the (first?) readings while it was still trying to connect?

If so, then it can be explained by the fact the Vcc is not read during the WiFi connect, as reading it may affect WiFi performance and vice verse.

I told you before the loop count is not a really useful measure for detecting the CPU load.
The loop count is determined by the scheduled jobs in the scheduler.
If nothing has to be done, the loop count can become quite low.
If you have the "eco" mode enabled, its count can even become lower.

v-a-d-e-r commented:
Now I merged this:
[WiFi] Convert bit-wise operations to use bitSet/bitClear/bitRead 
and it looks a bit better. And the Vcc is also correct! Strange thing....
Next test is with core 2.7.2, that was not good at this morning. Many MQTT disconnects and slow WiFi.

TD-er commented:
Usually I wait a few days before trying the new core versions, especially if they include a new toolchain..

But if you build the 'stage' build, you would actually have the same code as now on the master branch of esp8266/Arduino.

v-a-d-e-r commented:
I have flashed a new FW this minute with your changes and core 2.7.2 and let it run over night now. Will see what the log says tomorrow.... ;-)

TD-er commented:
Well I do have some test code pending with more logging.
But the more I look into it, it looks like the only way to know for sure what's happening is not to look at `WiFi.status()` or `WiFi.isConnected()` as they really sometimes make no sense.
We really should only look at our own internal state which is being handled via events.

v-a-d-e-r commented:
Wow, after 9 hours no crash (reboot), no lost WiFi! My config: ESPeasy custom, core 2.7.2, controllers 2 + 13 + 16 and 1 plugin (P001). But this is with my test node. Will now see what my "real" nodes do after flashing this FW...

TD-er commented:
Yep a lot of stability related fixes lately 
But this WiFi stuff keeps bogging me

v-a-d-e-r commented:
Ok, this was the last of my 18 nodes I flashed (all are different). This is a high risk to do this on daily needed nodes, but the only real test with usable results. But for now it looks not so bad.... :-)

v-a-d-e-r commented:
Ok, this was the last of my 18 nodes I flashed (all are different). This is a high risk to do this on daily needed nodes, but the only real test with usable results. But for now it looks not so bad.... :-)

TD-er commented:
OK, I have to make a decision now, how to tackle the WiFi issues.

Option 1:
Try to detect when WiFi status is out of sync (using events, timeouts and failed connect attempts)
Reset WiFi + try to reconnect until WiFi status is correct.

Option 2:
Ignore WiFi status and solely base WiFi true state on events.


What I observe here in my tests, is that I do not seem to miss any events.
So option 2 does seem to be the preferred way as it will result in much quicker response to WiFi state changes and thus less likely to crash.
For example sending UDP messages with incorrect WiFi state is likely to cause crashes.
Also the WiFi status may return WL_DISCONNECTED, even though it is actually connected and thus ready to be used. => improve of response time.

But some internals of libraries or the core may still rely on the `WiFi.status()` result, so we may run into issues which may be hard to reproduce or locate.

So what's the better option to do here?

v-a-d-e-r commented:
Is option 2 not a bit risky to ignore the WiFi status? BTW: At this time all my nodes seem to run stable with your last changes.

TD-er commented:
Well if the wifi status is incorrect, what should you do?
And the status can become incorrect during WiFi connect or when the node gets disconnected.

The tricky part here is that some nodes appear to have a higher chance of seeing this out of sync behavior compared to others.
I now have a node here on my desk which is very good at reproducing this behavior while my standard test nodes do connect and disconnect without issues.

v-a-d-e-r commented:
Hard to say. I have here 16 Sonoff nodes and 2 ESP-12F in use. All running without issues now. But I had not really strange WiFi issues in the past. Maybe it depends on the good router (FB7490)!?

TD-er commented:
I used to think that too, but since I now have one node that has significant more issues connecting to WiFi, compared to the other nodes, I don't think it is solely the router that makes the difference.

After all, it has been reported by users for ages, but I simply was unable to reproduce it.

v-a-d-e-r commented:
Or it depends on the stuff that is compiled within the firmware!?

TD-er commented:
Well it may also have some impact, but right now I do have 2 nodes here running the exact same binary and they do show different results.

So it may be some timing issue here or maybe a revision change of the ESP chip.
The one having the issues is a node I bought quite some time ago, which has a 1.3" OLED on board and the one on my breadboard has the same OLED chip but then connected via wires.
This makes it as similar as possible, but still there is a difference in behavior... exactly as being reported for ages by several users and until now I have not been able to reproduce it.
And I'm 100% sure now that there must be a bug in the Espressif SDK about the internal state as I'm also calling the SDK function directly. The same function as called in the esp8266/Arduino library.  See: https://github.com/esp8266/Arduino/issues/7432

This makes me quite happy, as I am now absolutely sure about what is going on here. Only problem now is what the next step should be.

v-a-d-e-r commented:
It's hard to believe that there is a change of the ESP chip design. I bought not all my nodes at the same time, so there would be also a difference. But they all act the same way....
That is the config how I compile. Maybe you see a difference to yours:
![20200708162815188](https://user-images.githubusercontent.com/36039152/86936122-e0fc0500-c13d-11ea-8f26-f3b816293e4d.jpg)

TD-er commented:
Only difference I can see right now is that you're building for QIO flash communication, which is faster.
But not supported by all boards.

v-a-d-e-r commented:
That's right. But I changed all flash chips to winbond 4M. And they can do QIO. :-)
My nodes are still running for over 11 hours now with your fixes. ;-)

uzi18 commented:
@TD-er not looked at code, but maybe if it possible to have runtime option to choose in default - actual behavior and alternative as only event based, so if someone will get issues we can suggest to try 2nd one ?

TD-er commented:
@uzi18 Well I attempt to get rid of very complex WiFi code and leaving both options is perhaps not the best way to make it less complex.

But I'll see what I can do. After all, it may end up with a function to detect the out of sync and a decision on what to do with it.

uzi18 commented:
@TD-er with such approach you can develop and test new code and drop old one later.
Ok will try to find a bit of time to understand your new code if you push it to any PR ;)
I'm very busy these days ;/

TD-er committed:
[WiFi] No longer rely on WiFi.isConnected()

TD-er committed:
[WiFi] Detect and work around missing events

The code is still way too complex, but at least it is working now on all tests I could throw at it.

TD-er committed:
Fix issue where task name / task var name was incorrect

TD-er committed:
Reduce minimal OTA build slightly

TD-er committed:
[WiFi] Small fixes for WiFi related code

TD-er committed:
[ESP32] Fix build error on ESP32

TD-er commented:
I made a new test build, would be great if it could be tested on a few nodes that used to have issues with connecting to WiFi

- [ESPEasy_ESP82xx_mega-20200703-39-PR_3148.zip](https://www.dropbox.com/s/kpwzx748jfv43cs/ESPEasy_ESP82xx_mega-20200703-39-PR_3148.zip?dl=0)
- [ESPEasy_ESP32_mega-20200703-39-PR_3148.zip](https://www.dropbox.com/s/1kk35m4f1yuh7vg/ESPEasy_ESP32_mega-20200703-39-PR_3148.zip?dl=0)

v-a-d-e-r commented:
I merged your last code and now the most nodes are not booting up after OTA flashing. Power cycling brings them up to life. This issue wasn't before. But I can not 100% say that I did not make a mistake... :-/

TD-er commented:
Are you sure you're using the same core?
In the esp8266/Arduino core there have been a number of changes regarding the boot sequence and XMC flash chips.

v-a-d-e-r commented:
I use core 2.7.2 and good winbond flash chips. Had never any issues with all my nodes. Ok, I can check the merged code a 3rd time....

TD-er commented:
Well if you use core 2.7.2 then that's already different from what I'm now using.
I did not yet change to core 2.7.2 as it caused a number of issues here which I still want to look into.

v-a-d-e-r commented:
Ah, ok. But I will be sure not to have a mistake somewhere I did myself. But 2.7.2 was running surprising stable before I merged your changes.

TD-er commented:
And is it running stable again after the (initial) power cycle?
Mainly the syslog (which could cause crashes) and strange WiFi reconnect loops etc. should have improved.

v-a-d-e-r commented:
They are not a long time alive now. For now it looks all normal.
BTW: Is there a reason not converting to the new bitRead()?

    if (!bitRead(wifiStatus, ESPEASY_WIFI_SERVICES_INITIALIZED)) { << NEW
      markWiFi_services_initialized();
    }
  }

  if (wifiStatus == ESPEASY_WIFI_DISCONNECTED) { << OLD ????
    #ifndef BUILD_NO_DEBUG

TD-er commented:
> They are not a long time alive now. For now it looks all normal.
> BTW: Is there a reason not converting to the new bitRead()?
> 
> ```
> if (!bitRead(wifiStatus, ESPEASY_WIFI_SERVICES_INITIALIZED)) { << NEW
>   markWiFi_services_initialized();
> }
> ```
> 
> }
> 
> if (wifiStatus == ESPEASY_WIFI_DISCONNECTED) { << OLD ????
> #ifndef BUILD_NO_DEBUG

Well comparing all flags to 0 is not really a shorter read using `bitRead` is it? ;)

v-a-d-e-r commented:
Ok, the 6 nodes I flashed yesterday survived the night without any issues... :-)

TD-er commented:
Maybe you can also let the node force to reconnect to WiFi and/or kick a node from the access point (if possible, or restart the AP)

TD-er committed:
Merge branch 'mega' into bugfix/wifi_reset

v-a-d-e-r commented:
Ooops, CPU load is rising from 10 to 80% after I disabled WLAN at the router. After connection is done it goes back to 10

TD-er commented:
CPU load (as it is computed) will indeed seem to go up.
But there's a simple explanation for it.
During WiFi connecting phase, there are some calls to `delay(10)` to give the WiFi part of the code ample amount of time to process all rather time critical steps.
And since this limits the number of loops that can be run, it will appear the load spikes.

The only way to make this look less (remember the load computation is based on the time spent in loop calls that don't have a scheduled job to run) is to schedule the wifi connect calls too.
Right now, it is still a fixed delay time, but it could be made a bit more dynamic by actually schedule the wifi maintenance.

v-a-d-e-r commented:
When no sensor data are lost with that high load it doesn't care to me. ;-)

TD-er commented:
> When no sensor data are lost with that high load it doesn't care to me. ;-)

It is only lost if the controller buffer is full and it cannot offload it to the other end of the controller because there is no network connection.
If the load gets to 100% (or very close), then there may be issues where the node cannot perform its scheduled jobs or react in due time.

v-a-d-e-r commented:
Under normal circumstances WLAN should be up and the nodes are connected. Most important is, that they do not crash and reboot with a lost WiFi...

TD-er committed:
[WiFi] Clear event callback on ESP32 + correct order of log/disconnect

TD-er committed:
[PIO] revert custom_beta env to stage core lib

TD-er committed:
[PIO ESP32] Update to latest Platformio espressif32@1.12.4

TD-er commented:
Just added some patches mainly for ESP32 and if you're logging to syslog.

But I also looked into the 'lockup' you mentioned @v-a-d-e-r 
Maybe it is something related to the serial port chip on the unit.

On boards with the CP210x chips when I had VS code to read the serial log and close VS code, the node hangs.
There should be no interaction with this serial to USB chip, but apparently there is.
These boards always have had the strange behavior that when you connect to them to read the logs they would trigger a reset.
Maybe there's something related here which triggers the reset too short so it will hang?

v-a-d-e-r commented:
I have no ESP32 in use. It is related to all nodes. Only 2 of 18 use serial port.

TD-er commented:
> I have no ESP32 in use. It is related to all nodes. Only 2 of 18 use serial port.

The USB to serial thingy I've only seen on the ESP8266 boards.
ESP32 doesn't seem to suffer from this
The latest fixes are mainly ESP32 related, so if you don't have one of those, I doubt you will see a lot of changes with the last commits.

v-a-d-e-r commented:
Also available for ESP32 with USB. See here....
https://www.ebay.de/itm/ESP-WROOM-32-Modul-Entwicklungsboard-WiFi-Bluetooth-BT-BLE-MCU-Dual-Core-CPU/164098926277?hash=item26350ee6c5:g:-GcAAOSwO5BeVxtw

TD-er commented:
The USB chip is present on almost all boards.
What I meant was the "device hangs when closing IDE while serial terminal present" issue.
That behavior I've only seen on ESP8266, not on ESP32.

By the way, what an exorbitant high price for such a module.

v-a-d-e-r commented:
That's right. I took the first hit I found, because I needed only the link ;-)

TD-er committed:
[ESPEasy p2p] Reload UDP port settings when changed

No need to reboot after changing UDP port setting.

TD-er committed:
[UDP] Reduce max UDP packet size

The max. UDP packet size could allow for a large memory allocation when a large packet was received.
No matter if this was a valid packet to process anyway.

TD-er committed:
[NTP] After timeout or receiving invalid packet, port could be held

This could lead to extra memory usage until the next (valid) NTP packet was processed.
Also the UDP port could be held.

TD-er committed:
[Syslog] Add custom port to the syslog IP (#1258)

Fixes: #1258

TD-er committed:
[Docs] Update docs related to p2p UDP port no longer need reboot

TD-er committed:
[UDP p2p] Fix crash on ESP32 where UDP is started before connected

TD-er committed:
[ESP32] Recognize last boot state as Deepsleep (#3155)

Fixes: #3155

For the ESP8266, we keep things like the last boot cause in RTC memory.
Currently that's not supported on ESP32 (in our code) so the last boot cause is never read and stored.

TD-er committed:
[TimeZone] Move timezone configuration line to location settings (#3141)

Fixes: #3141

TD-er committed:
[Docs] Document Tools => Advanced Settings

TD-er committed:
[WiFi] Copy DNS servers from DHCP when using fixed last octet (#3111)
